### ISSUE TITLE
request: return a function from modSearch

### ISSUE THREAD

**Issue by crazytyper** [svaarala/duktape#1985]
We are using duktape with CommonJS modules.

In our environment we frequently require and run the same large modules.

In order to improve performance we thought we could pre-compile our modules using `duk_dump_function` &`duk_load_function`.

This would be pretty easy if `modSearch` would support returning a function.

However `modSearch` currently only supports returning a string (containing the source code) or - if undefined is returned - it expects `module.exports` to be fully populated.

This would require us to create the CommonJS environment ourselves, which would be a shame as https://github.com/svaarala/duktape/blob/master/extras/module-duktape/duk_module_duktape.c already contains the code to do that.

Long story short:
```javascript
function myModSearch(...) {
  // get the pre-compiled module from a cache
  duk_load_function(...);
  return 1; // would be cool if duktape would proceed by initializing the module.
}
```

**Comment by svaarala** [comment]
Agreed, this is actually listed as a future work for modules. The minor small problem is that the module body is automatically wrapped into a function body in the calling module code and modSearch has no access to that wrapped function.

If one were to byte compile that wrapped function, and then load it back, it should be otherwise straightforward to get the module load correctly from dumped code.

**Comment by crazytyper** [comment]
Thank you for the quick and elaborate clarification.

I see there are a lot of details to sort out for a generic solution.
Some of them (like restoring native state) give me goosebumps ;)

As the roadmap is already neatly laid out in `doc/bytecode.rst` I would argue that this request can be closed.

For what it's worth ...
This is (a sketch of) the solution I currently use in my project:

```golang
import "github.com/olebedev/go-duktape"

// proper error handling skipped for brevity.

func modSearch(js *duktape.Context) int {
    // duk_tape stack layout:
    // 0: module id
    // 1: require function
    // 2: exports
    // 3: module
    const argModuleID = 0
    const argRequire = 1
    const argExports = 2
    const argModule = 3

    id := js.GetString(argModuleID)

    byteCode := getModuleByteCodeFromCache(id) // <-- my magic module cache is here

    if byteCode == nil {
        // cache miss! compile from source
        sourceCode := getModuleSource(id)

        // wrap source into CommonJS envelope
        wrappedSourceCode := "(function(require,exports,module){" + sourceCode + "\n})"
        
        // compile
        js.PushString(wrappedSourceCode)
        js.PushString(id + ".js")
        js.Pcompile(0)

        // dump byte code onto the stack
        js.DumpFunction()

        // put byte code into cache
        buffer, size = js.GetBuffer(-1)
        byteCode = bufferToBytes(buffer, size)
        js.Pop()
        putModuleByteCodeIntoCache(id, byteCode)
    }

    // load the byte code onto the stack
    buffer := js.PushBuffer(len(byteCode), false)
    bytesToBuffer(buffer, byteCode)
    js.LoadFunction()

    // eval'd function wrapper (not called yet)
    if rc := js.Pcall(0); rc != duktape.ExecSuccess {
        return duktape.ErrRetError
    }

    // call the function wrapper to initialize the module
    js.Dup(argExports) // <-- binding "this" to exports
    js.Dup(argRequire)
    js.GetPropString(argModule, "exports")
    js.Dup(argModule)
    if rc := js.PcallMethod(3); rc != duktape.ExecSuccess {
        return duktape.ErrRetError
    }
    return 0
}
```
